### PR TITLE
Update pre issue access token password grant test.

### PR DESCRIPTION
**Pull request by malithie**
Relates to https://github.com/wso2/product-is/issues/20842

This PR,

- Validates request initiated for the external endpoint that implements pre issue access token action. Several model classes are introduced with this.
- Refactor actions mock server

Kindly note this PR adds minimal refactoring and improvements to make sure the respective functionality is covered. This can be improved more as other cases are covered

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java [hunk 1]
```diff
@@ -122,24 +129,47 @@ public class PreIssueAccessTokenPasswordGrantTestCase extends ActionsBaseTestCas
...
+    private ActionsMockServer actionsMockServer;
+
+    @Factory(dataProvider = "testExecutionContextProvider")
+    public PreIssueAccessTokenPasswordGrantTestCase(TestUserMode testUserMode) {
+
+        this.userMode = testUserMode;
+        this.tenantId = testUserMode == TestUserMode.SUPER_TENANT_USER ? "-1234" : "1";
+    }
+
+    @DataProvider(name = "testExecutionContextProvider", parallel = true)
```
Comment (ashanthamara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java:165]: Shall we move this to ActionBaseTestCase class and add the dataprovider to the the ClientCredentials test class as well ?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java [hunk 2]
```diff
@@ -149,7 +179,7 @@ public class PreIssueAccessTokenPasswordGrantTestCase extends ActionsBaseTestCas
     @BeforeClass(alwaysRun = true)
     public void testInit() throws Exception {
 
-        super.init(TestUserMode.TENANT_USER);
+        super.init(userMode);
```
Comment (ashanthamara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java:182]: Let's move this to the constructor.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java [hunk 3]
```diff
@@ -177,32 +207,52 @@ protected boolean isRedirectable(String method) {
...
+        requestedScopes.addAll(customScopes);
+
         actionId = createPreIssueAccessTokenAction();
+
+        actionsMockServer = new ActionsMockServer();
+        actionsMockServer.startServer();
+        actionsMockServer.setupStub(MOCK_SERVER_ENDPOINT_RESOURCE_PATH,
+                "Basic " + getBase64EncodedString("test", "test"),
+                FileUtils.readFileInClassPathAsString("actions/response/pre-issue-access-token-response.json"));
+
```
Comment (ashanthamara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java:236]: Remove extra line

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java [hunk 4]
```diff
@@ -250,7 +288,21 @@ public void testGetAccessTokenWithPasswordGrant() throws Exception {
...
-    @Test(groups = "wso2.is", description = "Verify the presence of the updated scopes in the access token")
+    @Test(groups = "wso2.is", dependsOnMethods = "testGetAccessTokenWithPasswordGrant", description =
+            "Verify the pre issue access token action request")
+    public void testPreIssueAccessTokenActionRequest() throws Exception {
+
+        String actualRequestPayload = actionsMockServer.getReceivedRequestPayload(MOCK_SERVER_ENDPOINT_RESOURCE_PATH);
+        PreIssueAccessTokenActionRequest actualRequest =
+                new ObjectMapper().readValue(actualRequestPayload, PreIssueAccessTokenActionRequest.class);
+
+        PreIssueAccessTokenActionRequest expectedRequest = getRequest();
```
Comment (ashanthamara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java:299]: Shouldn't we validate this in other grant types as well ? If so let's move the `getRequest()` to the ActionBaseTestCase class

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/model/ActionType.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package org.wso2.identity.integration.test.actions.model;
+
+/**
+ * This class models the Action Type.
+ */
+public enum ActionType {
+    PRE_ISSUE_ACCESS_TOKEN,
+    AUTHENTICATION
```
Comment (ashanthamara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/model/ActionType.java:26]: Let's remove `AUTHENTICATION` for now and add it when the authentication action type test cases are introduced.

File: modules/integration/tests-integration/tests-backend/src/test/resources/actions/response/pre-issue-access-token-response.json [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    {
+      "op": "remove",
+      "path": "/accessToken/scopes/6"
+    },
+    {
+      "op": "remove",
+      "path": "/accessToken/claims/aud/-"
+    }
+  ]
+}
```
[the remark this answers is not in the document]
Comment (malithie) [reply modules/integration/tests-integration/tests-backend/src/test/resources/actions/response/pre-issue-access-token-response.json:94]: This is a particular response. Not a code file. Let's not add lines here.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java (hunk 2 above)
Comment (malithie) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java:182]: We should not move this to the constructor. Constructor just instantiates the test class. This is the exact test case init.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java (hunk 4 above)
Comment (malithie) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java:299]: Request can differ based on grant types. So let's keep this here. We might need to find what's common and fix accordingly

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java [hunk 7]
```diff
@@ -328,10 +454,10 @@ public void testTokenExpiresInClaimReplaceOperation() throws Exception {
     private String createPreIssueAccessTokenAction() {
 
         AuthenticationType authenticationType = new AuthenticationType();
-        authenticationType.setType(AuthenticationType.TypeEnum.BASIC); // todo handle mock server authorization
+        authenticationType.setType(AuthenticationType.TypeEnum.BASIC);
         Map<String, Object> authProperties = new HashMap<>();
-        authProperties.put(USERNAME_PROPERTY, TEST_USER);
-        authProperties.put(PASSWORD_PROPERTY, ADMIN_WSO2);
+        authProperties.put(USERNAME_PROPERTY, "test");
```
[the remark this answers is not in the document]
Comment (malithie) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenPasswordGrantTestCase.java]: Fixed

**Review by ashanthamara — APPROVED** [review]

**Pull request merged**